### ISSUE TITLE
Add mechanism for optional backend functions

### ISSUE THREAD

**Issue by josevalim** [elixir-nx/nx#362]
While we provide built-in implementations for inverse, tile, etc. This has the benefit that we get `grad` of these functions for free. However, we also want to allow backends to provide their own implementation for performance reasons.

This may seem straight-forward at first but the biggest issue is related to how optional backend function ties in to Expr/grad. We will need Expr to implement all optional callbacks while keeping the default implementation for grad.

**Comment by seanmor5** [comment]
I propose adding a `custom_call/2` metadata node or similar which looks something like:

`custom_call(:target, fallback_impl_fn)`

Then compilers and backends can register custom calls for specific targets, during evaluation and compilation those special targets are dispatched to the registered custom call implementation. In the backward pass it is just metadata so the gradient is still valid.

This would be useful in `defn` as well. For example in Axon some functions or layers might have specialized implementations. BatchNorm for example has an XLA specialized implementation which is faster than the one that is generated from the code. Torchx would benefit as well because things like softmax and other activations would not need to fall back on the evaluator
